### ISSUE TITLE
Contents of DroidGuard binary blob

### ISSUE THREAD

**Issue by doug-leith** [microg/android_packages_apps_GmsCore#1139]
Has anyone managed to figure out what the content of the binary blob returned by DroidGuard (and sent to Google servers) actually is?   The stuff I find online is mostly quite old now, and v limited.  Thanks.

**Comment by ArchangeGabriel** [comment]
No, we don’t know. You would have to do reverse-engineering for that. And if we knew, we could probably circumvent SafetyNet.

**Comment by doug-leith** [comment]
Thanks.  I’m interested in the privacy implications, if any, i.e. does it contain any device/user identifiers.   Reverse engineering DroidGuard is not so easy of course, but surely not impossible either.

**Comment by ArchangeGabriel** [comment]
> I’m interested in the privacy implications, if any, i.e. does it contain any device/user identifiers.

Likely, but we don’t know for sure.

**Comment by Romern** [comment]
Some insight about the blob itself can be found in this article: https://habr.com/en/post/446790/  
It depends on the specific request itself, but in most cases I have seen your Google device ID along with the package signature will be used to sign requests, e.g. in https://github.com/microg/android_packages_apps_GmsCore/blob/master/play-services-core/src/main/java/org/microg/gms/snet/SafetyNetClientServiceImpl.java . Afaik DroidGuard itself does not send the request to the server, but is only used by SafetyNet and ther services (e.g. FirebaseAuth) to sign the requests.

**Comment by ArchangeGabriel** [comment]
> Some insight about the blob itself can be found in this article: https://habr.com/en/post/446790/

Very interesting reading, thanks!

**Comment by doug-leith** [comment]
Thanks, yes I saw that already and sent them an email too (am hoping I’ll not have to redo that painful reverse engineering!)

**Comment by squeaktoy** [comment]
https://www.romainthomas.fr/publication/22-sstic-blackhat-droidguard-safetynet/
Here is a whitepaper about the way DroidGuard works.
